core/src/main/java/org/elasticsearch/threadpool/ThreadPool.java

```diff
@@ -971,4 +980,147 @@ public void close() throws IOException {
     public ThreadContext getThreadContext() {
         return threadContext;
     }
+
+    static class ReschedulingRunnable implements Runnable {
+
+        private static final int STOP = 0;
+        private static final int RUN = 1;
+
+        private final CountDownLatch runLatch = new CountDownLatch(1);
+        private final CancellableScheduledFuture future = new CancellableScheduledFuture();
+        private final Runnable runnable;
+        private final TimeValue interval;
+        private final String name;
+        private final ThreadPool threadPool;
+
+        private volatile Throwable thrownException = null;
+
+        ReschedulingRunnable(Runnable runnable, TimeValue interval, String name, ThreadPool threadPool) {
+            this.runnable = runnable;
+            this.interval = interval;
+            this.name = name;
+            this.threadPool = threadPool;
+        }
+
+        ScheduledFuture<?> scheduleFirstRun() {
+            if (runLatch.getCount() == STOP) {
+                throw new IllegalStateException("this runnable should not be reused after cancelling");
+            }
+            threadPool.schedule(interval, name, this);
+            future.updateNextTime(System.nanoTime());
+            return future;
+        }
+
+        @Override
+        public void run() {
+            assert future != null;
+            try {
+                if (runLatch.getCount() == RUN) {
+                    // if a runnable throws, the default behavior of ScheduledThreadPoolExecutor is to stop execution
+                    runnable.run();
+                }
+
+                if (runLatch.getCount() == RUN) {
+                    threadPool.schedule(interval, name, this);
+                    future.updateNextTime(System.nanoTime());
+                }
+            } catch (Throwable t) {
+                // if anything throws move to STOP
+                thrownException = t;
+                runLatch.countDown();
+                throw t;
+            }
+        }
+
+        class CancellableScheduledFuture implements ScheduledFuture<Void> {
```

Comment:
I'm ok with simplifying if we do not return a ScheduledFuture as it feels wrong/misleading to return a class where we implement a single method and throw an exception for the other methods. 

Instead, we can return our own class that can be used to cancel and that is it. We could call it `CancellableTask` or something and just implement a `cancel` method. What do you think about that?